Add Google colab example

shadiakiki1986 opened the pull request:
Based on feedback from https://github.com/rlworkgroup/garage/issues/470

ryanjulian commented:
@shadiakiki1986 this is awesome and we'd love to merge it.

There have been a couple recent API changes which make this need an update (particularly the use of `LocalRunner`. Please take a look at the updated examples in `examples/tf/` for the new API.

Let us know if you need help!

shadiakiki1986 commented:
Hello @ryanjulian 
I modified the notebook to use the current garage API.
The original notebook was for gym `CartPole`.
I added another notebook for a custom user gym environment too.
Let me know what you think :clinking_glasses:

shadiakiki1986 commented:
Btw I also needed to add a new `scipts/setup_colab.sh` because the `setup_linux.sh` doesn't work out-of-the-box for colab.research.google.com (e.g. it creates a new conda environment)

shadiakiki1986 commented:
Hey there @ryanjulian . Any news on this PR?

ryanjulian commented:
@gonzaiva can you take a look and make sure you can replicate @shadiakiki1986 's setup?

gonzaiva commented:
Hi @shadiakiki1986,

Thanks for helping us with this feature. I just tried your script in this little notebook I made:
https://colab.research.google.com/gist/gonzaiva/59392151ade12a4ee6c7fd153ee0d1bb/test_garage_co.ipynb
It worked, but I will add some comments in your script.

Also, do you know if there's a way to build a remote session in google colab with garage preloaded so we don't have to do the installation step?
That would make things much simpler.

@ryanjulian to get some logs, `run_experiment` adds the outputs to the logger, but since we cannot use `run_experiment`, that step has to be done manually as shown in the notebook I made. Should we add some lines of code in the setup call of the LocalRunner to add `stdout` in case no output has been added?

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+      --mjkey=*)
+        _arg_mjkey="${_key##--mjkey=}"
+        ;;
+      --no-modify-bashrc|--modify-bashrc)
```
Besides `mjkey` and `h`, all the other arguments are not used here, so it's better to remove all the code associated to them. The code for the arguments in this script was generated with [argbash](https://argbash.io/generate), so I would recommend generating them again with only the `mjkey` and `h` arguments.

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+# Install command "file" if not found
+echo "Install command "file" if not found"
+if [[ -z "`command -v file`" ]]; then
+  apt-get install -y -q file libmagic-dev
```
What is libmagic-dev used for?

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+      --mjkey=*)
+        _arg_mjkey="${_key##--mjkey=}"
+        ;;
+      --no-modify-bashrc|--modify-bashrc)
```
Besides mjkey and h, all the other arguments are not used here, so it's better to remove all the code associated to them. The code for the arguments in this script was generated with argbash, so I would recommend generating them again with only the mjkey and h arguments.

gonzaiva replied:
Ups, it seems `gpu` can be used in colab (the future is now):
https://colab.research.google.com/notebooks/gpu.ipynb
So, also make sure to include the `gpu` argument please.

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+  _PRINT_HELP=yes die "${dir_err_txt}" 1
+fi
+
+# Install command "file" if not found
```
By running `!bash file -h` in a fresh notebook runtime in colab, I confirmed that `file` is already installed, so why do we need to install it again?

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+# Required for mujoco_py: libglew-dev, patchelf, libosmesa6-dev
+# Required for OpenAI gym: libpq-dev, ffmpeg, libjpeg-dev, swig, libsdl2-dev
+# Required for OpenAI baselines: libopenmpi-dev, openmpi-bin
+>&2 echo "start of setup_colab.sh"
```
Remove

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+
+}
+
+>&2 echo "end of setup_colab.sh"
```
Remove

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+  # https://github.com/openai/mujoco-py/issues/66
+  # pip install git+git://github.com/openai/mujoco-py.git@1.50.1.1 # requires mujoco activation key
+  MUJOCO_150="$(mktemp -d)/mujoco_150"
+  git clone --branch 1.50.1.1 --depth 1 https://github.com/openai/mujoco-py.git $MUJOCO_150
```
`setup.py` inside garage already includes `mujoco-py`, so once you execute `pip install -e [all].`, that `mujoco-py` installation will override this one. Remove.

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+  
+  
+  # install dm-control related to mujoco 1.50
+  # pip install -v git+git://github.com/deepmind/dm_control.git@mujoco1.50
```
Remove

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+  
+  # install dm-control related to mujoco 1.50
+  # pip install -v git+git://github.com/deepmind/dm_control.git@mujoco1.50
+  pip install git+git://github.com/deepmind/dm_control.git@92f9913013face0468442cd0964d5973ea2089ea
```
Same comment as above related to `mujoco-py`,  this package is already included in `setup.py`. Remove.

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+  pip install git+git://github.com/deepmind/dm_control.git@92f9913013face0468442cd0964d5973ea2089ea
+
+  # equivalent of conda env -f environment.yml
+  pip install numpy==1.14.5
```
Same comment as above related to `mujoco-py`, this package is already included in `setup.py`. Remove.

gonzaiva commented on scripts/setup_colab.sh:
```diff
@@ -0,0 +1,243 @@
+  # equivalent of conda env -f environment.yml
+  pip install numpy==1.14.5
+  pip install -e .[all]
+  pip install -e .[dev]
```
The `dev` installation contains the `all` installation, so I don't see the purpose of having both. Moreover, I don't think the packages in the `dev` installation would be used in python notebooks, since most of them are lint and unit tests tools that are used when submitting code to garage, so `pip install -e .[all]` should be sufficient for the notebook purposes. Remove `pip install -e .[dev]`

ryanjulian commented:
@gonzaiva i don't think all of these adjustments are really necessary for an example code. Did you verify that it works on colab?

As long as it works, I'm happy.

ryanjulian commented:
@gonzaiva no -- LocalRunner should not setup the logger. That is the job of the launcher or experiment runner.

Looking at your notebook, it looks fine to me.

gonzaiva commented:
@ryanjulian, yes, I verified it, check the notebook at:
https://colab.research.google.com/gist/gonzaiva/59392151ade12a4ee6c7fd153ee0d1bb/test_garage_co.ipynb

ryanjulian commented:
@shadiakiki1986 please rebase onto master and then I will merge this.

Thanks again for the patch! Sorry it took so long.

shadiakiki1986 commented:
@ryanjulian rebase done :+1:
